devdocs/PIPELINE_REFACTORING_PLAN.md

```diff
@@ -2,276 +2,481 @@
-  - Move processing logic out of `create_ummg()` into dedicated processors
-  - Create `ummg_builder.py`, `cnm_builder.py` for template application
-  - Move template population out of operations into builders
-  - Create `file_writer.py`, `aws_writer.py` for final phase I/O
-  - Extract file writing and AWS operations from pipeline operations
-  - Implement batch operations for efficiency
-- **Risk**: High - core business logic changes with medium risk for isolated template and I/O logic
-
-### Story 5: Enhanced Pipeline Management
-- **Goal**: Create comprehensive pipeline orchestration with monitoring and error handling
-- **Tasks**:
-  - Create `PipelineOrchestrator` class to manage all phases
-  - Implement proper error handling mechanisms
-  - Add pipeline monitoring and metrics collection
-  - Replace current logging with structured result collection
-  - Add detailed success/failure reporting with granule-level tracking
-  - Implement pipeline performance metrics and timing
-- **Risk**: Low - new functionality that enhances existing capabilities
-
-## Implementation Strategy
-
-### Backward Compatibility
-- Each story maintains existing CLI interface
-- No breaking changes to configuration format
-- Existing behavior preserved throughout transition
-
-### Incremental Testing
-- Add comprehensive tests for each extracted component
-- Goal of test coverage above 90% from refactoring
-- Add integration tests for new pipeline phases
-
-### Feature Flags
-- Use configuration to toggle between old/new implementations during transition
-- Allow gradual rollout of new pipeline components
-- Provide fallback mechanisms for critical operations
+Key architectural improvements:
+- **Complete separation** between pure computation (Specification Phase) and side effects (Execution Phase)
+- **Extensible reader system** allowing custom readers for any file format
+- **Strategy pattern** for geometry and temporal processing based on CMR metadata
+- **Immutable state** threading through all pipeline operations
+- **Optimization opportunities** in the execution phase (parallel, batch, etc.)
+- **Comprehensive ledger** tracking all operations and errors
+
+### Implementation Priorities
+
+| Priority | Implement Phases | Key Benefits |
+|----------|------------------|--------------|
+| Foundation | Phases 1-3 | Extracted strategies + CMR reader |
+| Core Refactor | Phases 1-4 | Above + pipeline state architecture |
+| Extensibility | Phases 1-6 | Above + custom reader support |
+| Full Refactor | Phases 1-7 | Above + consistent reader architecture |
+| Complete | All 8 phases | Above + pure functional pipeline |
+
+### Critical Success Factors
+
+1. **Incremental Extraction**: Extract existing logic rather than rewriting
+2. **Maintain Compatibility**: Existing pipeline continues to work
+3. **Test Each Phase**: Ensure extracted components work identically
+4. **Feature Flags**: Toggle between old and new implementations
+
+## Benefits of This Approach
+
```

Comment:
This section feels like a repeat of the earlier list of benefits. The "table of contents" idea may help expose the redundancy if I'm not making it up.